contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// This is the address where the burned base fee go. We use this to send the amount of
+// burned tokens back to the source chain.
+address constant BURNED_TX_FEES_ADDRESS = 0x0100000000000000000000000000000000000000;
+// Designated Blackhole Address. This is where we burn token before sending an unlock
+// message to the source chain. Different from the burned tx fee address so we can
+// track each separately.
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000001;
+// Precompiled Warp address
+address constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable sourceContractAddress;
+    // We will not mint the first `tokenReserve` tokens sent to this subnet.
+    // This should be constructed to match the initial token supply of this subnet.
+    // This will mean we will not mint tokens until the source contact is collateralized.
+    uint256 public tokenReserve;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    // TODO we probably want to add the original token supply from this chain to the constructor.
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address sourceContractAddress_,
+        uint256 tokenReserve_
+    ) {
+        currentBlockchainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "Invalid Teleporter Messenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(sourceBlockchainID_ != bytes32(0), "Invalid Source Chain ID");
+        require(
+            sourceBlockchainID_ != currentBlockchainID,
+            "Cannot Bridge With Same Blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            sourceContractAddress_ != address(0),
+            "Invalid Source Contract Address"
+        );
+        sourceContractAddress = sourceContractAddress_;
+
+        tokenReserve = tokenReserve_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
```

Comment:
can we rename to originChainID and originSenderAddress, or sourceChainID and sourceSenderAddress? Maybe the latter to match `NativeTokenSource` naming

Reply:
Anyone can send a teleporter message to this function, so all we know are they there are the `sender` of the message.